### ISSUE TITLE
Appending rows to a large table.

### ISSUE THREAD

**Issue by jbucaran** [hyperapp/hyperapp#216]
[New benchmarks](https://rawgit.com/krausest/js-framework-benchmark/master/webdriver-ts/table.html) show we've improved a lot since hyperapp was first introduced, but we're still performing comparatively poor in @krausest / js-framework-benchmark: **Append rows to large table**: Duration for adding 1000 rows on a table of 10,000 rows.

I am not trying to win these benchmarks, but of all the things, I still can't figure out why what takes to beat this particular test.

![ha](https://cloud.githubusercontent.com/assets/56996/26433950/8fe91dea-4140-11e7-9c88-de6f40ca77cb.png)

@leooniya's domvm is particularly good on this one, maybe you can share a couple of secrets?

**Comment by leeoniya** [comment]
try to avoid recreating and/or patching the full unchanged 10k vtree via an eager diff/sCU that reuses old vnodes and essentially noops the extra work.

**Comment by jbucaran** [comment]
Focus has shifted to introducing rAF and vnode recycling. Closing in favor of #244.

**Comment by yelouafi** [comment]
@jbucaran I might be wrong but I suspect raf or recycling will help solving this issue. 

I think the issue is [here](https://github.com/hyperapp/hyperapp/blob/master/src/app.js#L211-L221). You're building the keymap too early, with a 10k table, you'll build an object with 10k which takes some time (even if in this case it is not necessary). 

I would suggest running some preoptimization passes to trim common prefixes/suffixes (compare by index and patch first/last matching elements) and then only diff using a map for the rest (starting from the first until the last non matched elements). In the above case a map is not needed at all, after trimming the common prefix (the 1st 10k rows unchanged) since the old sequence is empty, all we have to do is to append the remaining in the new sequences. You can also check my comment here (https://github.com/developit/preact/issues/725#issuecomment-313954428) for more details

**Comment by jbucaran** [comment]
> @yelouafi I might be wrong but I suspect raf or recycling will help solving this issue.

Just to clarify, did you mean to say perhaps they will _not_ help? 

> I would suggest running some preoptimization passes to trim common prefixes/suffixes

What is a prefix/suffic?

**Comment by leeoniya** [comment]
> What is a prefix/suffix?

https://github.com/localvoid/kivi/blob/master/lib/vnode.ts#L1288

**Comment by jbucaran** [comment]
The comment alone is bigger than hyperapp. 😆

**Comment by leeoniya** [comment]
but it minifies to 0, even before gzip!

**Comment by yelouafi** [comment]
>Just to clarify, did you mean to say perhaps they will not help?

Yeah, sorry for the typo.

>What is a prefix/suffix?

Those are the sequences of elements that are the same at the start (prefix) and at the end (suffix) in the old and the new sequences. For example

old sequence :   `A B C D E`
new sequence :  `A B x y C D E`

In this example 2 elements `x` & `y` were inserted in the middle. `[A,B]` is the common prefix and `[C,D,E]` is the common suffix. We can identify those with a simple loop

```js
newStart = 0
newEnd = newChildren.length
oldStart = 0
oldEnd = oldChildren.length


while(newStart <= newEnd && oldStart <= oldEnd) {
  // strip common prefix
  if(haveSameTypeAndKey(newChildren[newStart], oldChildren[oldStart]) {
    patch(newChildren[newStart], oldChildren[oldStart])
    newStart++; oldStart++
  }
  else if(haveSameTypeAndKey(newChildren[newEnd], oldChildren[oldEnd]) {
    patch(newChildren[newEnd], oldChildren[oldEnd])
    newEnd--; newEnd--
  }
  // can also do some other tricks like inverions etc
  // you can see https://github.com/yelouafi/petit-dom/blob/master/src/vdom.js#L141
}

/*
if old seq is empty -> insert new elements
example
  old = 'ABCD'
  new = 'ABxyCD'
  after prefix/suffix trimming; old = '' and new = 'xy'
*/
if(oldStart > oldEnd) {
  while(newStart <= newEnd)
    insertBefore(newChildren[newStart++], oldChidlren[oldStart])
}

/*
if new seq is empty -> remove old elements
example
  old = 'ABxyCD'
  new = 'ABCD'
  after prefix/suffix trimming; old = 'xy' and new = ''
*/
else if(newSart > newEnd) {
  while(oldStart<= oldEnd)
    remove(oldChidren[oldStart++])
}

else {
  // continue with map based diff but only with elements [newStart...newEnd] & [oldStart..oldEnd]
}
```

the prefix/suffix trimming will cover most of the cases (like in the above example, where the prefix will be the first 10k rows) and you wont need event to do a map based diff which is slower (construction of the map, lookup ...)

**Comment by jbucaran** [comment]
@yelouafi I want to make sure we are on the same page: in your example, AB C and D are keyed nodes and x, y un-keyed nodes. Is that so?

**Comment by yelouafi** [comment]
Trimming will work for both keyed and non keyed nodes. For non keyed nodes the comparaison will return true for nodes with the same type

**Comment by jbucaran** [comment]
Right, I think I meant to ask: why are some of the letters uppercase/lowercase?

**Comment by jorgebucaran** [comment]
Resolved in #499
